### ISSUE TITLE
Elaborate all operationally relevant information in the type checker

### ISSUE THREAD

**Issue by nomeata** [dfinity/motoko#53]
In a simple world, a type system tells you which programs will not fail, but do not affect the operational behavior. It is possible to ignore the types, and still run or compile the program, i.e. all operational information is present in the terms.

But in more complex languages, like ours, there are some operationally relevant details not manifest in the source term AST, but are filled in by type inference: E.g. whether `(-)` needs to prevent creating negative numbers or not; which number types to deal with; how to compile a function call etc.

So I would suggest that we extend the term AST with `foo option` fields that are always `None` after parsing, but are filled in by type inference with enough information that one can run or compile the programs without looking at the type annotations. E.g. for `Subtract` operation, it would specify the ground type and whether to trap on creating negative numbers.

This would avoid duplicating logic like this:
```ocaml
  | RelE (exp1, op, exp2) ->
    let t = T.lub Con.Env.empty (* both types are primitive *)
      (T.as_immut exp1.note.note_typ) (T.as_immut exp2.note.note_typ) in
```
which looks like it should be part of type inference (only), but currently needs to be included in the interpreter and (later) in the compiler.

TL;DR: The post type-checker AST should executable even after type annotations are erased.
